### ISSUE TITLE
obj.hasOwnProperty is not a function

### ISSUE THREAD

**Issue by ajainvivek** [domchristie/humps#35]
decamelizeKeys Object or Array of objects throws an error obj.hasOwnProperty is not a function in NodeJS environment.

**Comment by domchristie** [comment]
Are you able to give an example? Thanks.

**Comment by ajainvivek** [comment]
Example: 

```
const myArr = [{
 validTill: 'date obj',
 eventId: 'uniq id'
}]

const decamelizeKeys = humps.decamelizeKeys(myArr); // Throws error obj.hasOwnProperty is not a function

// OR

const decamelizeKeys = _.map(myArr, (item) => humps.decamelizeKeys(item)); // Throws error obj.hasOwnProperty is not a function
```

Environment: Node 6

It works fine on browser.

**Comment by domchristie** [comment]
Sorry I am unable to reproduce this. I have tested the following in Node 6.9.3 and it works as expected:

```
> var humps = require('./')
undefined
> const myArr = [{
...  validTill: 'date obj',
...  eventId: 'uniq id'
... }]
undefined
> 
> const decamelizeKeys = humps.decamelizeKeys(myArr);
undefined
> decamelizeKeys
[ { valid_till: 'date obj', event_id: 'uniq id' } ]
```

**Comment by lloydh** [comment]
Hi,

I came across this issue too trying to run decamelizeKeys on the output object produced by the Joi validator. It turns out that the .hasOwnProperty method gets stripped by Joi.

Since we can't rely on supplied objects having the hasOwnProperty method (or a modified version) I have created a [pull request](https://github.com/domchristie/humps/pull/36) to call hasOwnProperty directly from the Object prototype.

Lloyd

**Comment by lloydh** [comment]
Any chance of merging the fix and making a new release?

**Comment by levivoelz** [comment]
I'm running into the same issue, but in the browser. The PR above seems to fix it.

**Comment by khisakuni** [comment]
Any update on this? I'm running into this issue on the browser as well.

**Comment by domchristie** [comment]
`obj.hasOwnProperty` is [a widely recognised method to check object properties](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty), so I would like to understand why this is happening before merging. If anyone can provide a failing demo, I can look into it. Thanks.

**Comment by lloydh** [comment]
Yes, hasOwnProperty is the de facto way to check objects for properties, but unfortunately it's mutable so out in the wild it can't be relied on.

Here are a couple of examples of how you can't rely on the hasOwnProperty method of an object you haven't created.

```
const foo = Object.create(null);
foo; // {}
foo.bar = 'baz';
foo.hasOwnProperty('bar'); // TypeError: foo.hasOwnProperty is not a function

Object.prototype.hasOwnProperty.call(foo, 'bar'); // true
```

```
const foo2 = { bar: 'baz' };
foo2.hasOwnProperty = () => false;
foo2.hasOwnProperty('bar'); // false

Object.prototype.hasOwnProperty.call(foo2, 'bar'); // true
```

[More info...](http://eslint.org/docs/rules/no-prototype-builtins)

The comments above demonstrate this is more than just a hypothetical problem and I think that a library with no control over the objects that it receives should follow the principle:
_Be conservative in what you do, be liberal in what you accept from others_

**Comment by domchristie** [comment]
Perfect, thanks for the explanation! https://github.com/domchristie/humps/pull/36 has been merged and released in v2.0.1
